File: app/services/user_service.py
"""User service that uses MongoDB when available; otherwise uses in-memory store.
This mirrors behaviour from the Node.js user.service.js file where cognitoUserId is used as _id.
"""
from uuid import uuid4
from app.utils.mongodb import connect_to_database

_USERS = {}


def _db():
    return connect_to_database()


def create_user(data: dict) -> dict:
    db = _db()
    cognito_id = data.get("cognitoUserId") or data.get("cognitoUserId")
    if db:
        users = db.collection("users")
        if not cognito_id:
            raise ValueError("cognitoUserId is required")
        data["createdAt"] = None
        data["updatedAt"] = None
        existing = users.find_one({"_id": cognito_id})
        if existing:
            # update existing
            return update_user_by_cognito_id(cognito_id, data)
        # insert new
        payload = {"_id": cognito_id, **{k: v for k, v in data.items() if k != "cognitoUserId"}, "serie_subcribe": []}
        users.insert_one(payload)
        return {"_id": cognito_id, **payload}

    # fallback in-memory
    user_id = cognito_id or str(uuid4())
    user = {"_id": user_id, **data, "serie_subcribe": []}
    _USERS[user_id] = user
    return user


def get_user_by_id(user_id: str) -> dict:
    db = _db()
    if db:
        users = db.collection("users")
        return users.find_one({"_id": user_id})
    return _USERS.get(user_id)


def get_user_by_cognito_id(cognito_id: str) -> dict:
    return get_user_by_id(cognito_id)


def update_user(user_id: str, data: dict) -> dict:
    db = _db()
    if db:
        users = db.collection("users")
        for k in ("_id", "cognitoUserId", "createdAt"):
            data.pop(k, None)
        data["updatedAt"] = None
        result = users.find_one_and_update({"_id": user_id}, {"$set": data}, return_document=True)
        return result
    existing = _USERS.get(user_id)
    if not existing:
        return None
    existing.update(data)
    _USERS[user_id] = existing
    return existing


def update_user_by_cognito_id(cognito_id: str, data: dict):
    db = _db()
    if db:
        users = db.collection("users")
        for k in ("_id", "createdAt"):
            data.pop(k, None)
        data["updatedAt"] = None
        result = users.find_one_and_update({"_id": cognito_id}, {"$set": data}, return_document=True, upsert=True)
        return result
    # fallback
    existing = _USERS.get(cognito_id, {})
    existing.update(data)
    _USERS[cognito_id] = existing
    return existing

File: node_src/services/user.service.js
// No need for ObjectId since we're using string IDs (cognitoUserId)
const { connectToDatabase } = require("../utils/mongodb");
const { COLLECTION_NAME } = require("../models/user.model");

/**
 * User Service - Handles business logic for user-related operations
 */
class UserService {
  /**
   * Create a new user
   * @param {Object} userData - User data to save
   * @returns {Promise<Object>} Created user object
   */
  async createUser(userData) {
    const db = await connectToDatabase();
    const collection = db.collection(COLLECTION_NAME);

    // Check if cognitoUserId exists
    if (!userData.cognitoUserId) {
      throw new Error("cognitoUserId is required");
    }

    // Set timestamps
    userData.createdAt = new Date();
    userData.updatedAt = new Date();

    // Simply check if user already exists with cognitoUserId as _id
    const existingUser = await collection.findOne({
      _id: userData.cognitoUserId,
    });

    if (existingUser) {
      // If user exists, update information
      return this.updateUserByCognitoId(userData.cognitoUserId, userData);
    }

    // If user does not exist, create new using cognitoUserId as _id
    // We don't need to keep a separate cognitoUserId field since it's the same as _id
    const { cognitoUserId, ...restUserData } = userData;

    const result = await collection.insertOne({
      _id: cognitoUserId,
      ...restUserData,
      serie_subcribe: [],
    });

    if (!result.acknowledged) {
      throw new Error("Failed to create user");
    }

    return {
      _id: cognitoUserId,
      ...restUserData,
      serie_subcribe: [],
    };
  }

  /**
   * Get user by database ID
   * @param {string} userId - Database user ID (which should be cognitoUserId)
   * @returns {Promise<Object|null>} User object or null if not found
   */
  async getUserById(userId) {
    try {
      const db = await connectToDatabase();
      const collection = db.collection(COLLECTION_NAME);

      // Since we're using cognitoUserId as _id, we don't need ObjectId conversion
      return await collection.findOne({ _id: userId });
    } catch (error) {
      throw new Error(`Error fetching user: ${error.message}`);
    }
  }

  /**
   * Get user by Cognito user ID
   * @param {string} cognitoUserId - Cognito user ID
   * @returns {Promise<Object|null>} User object or null if not found
   */
  async getUserByCognitoId(cognitoUserId) {
    try {
      const db = await connectToDatabase();
      const collection = db.collection(COLLECTION_NAME);

      // Since cognitoUserId is now the document _id, we can query by _id directly
      return await collection.findOne({ _id: cognitoUserId });
    } catch (error) {
      throw new Error(`Error fetching user: ${error.message}`);
    }
  }

  /**
   * Update user information by database ID
   * @param {string} userId - Database user ID (which should be cognitoUserId)
   * @param {Object} userData - User data to update
   * @returns {Promise<Object|null>} Updated user object or null
   */
  async updateUser(userId, userData) {
    try {
      const db = await connectToDatabase();
      const collection = db.collection(COLLECTION_NAME);

      // Don't allow updating these fields directly
      delete userData._id;
      delete userData.cognitoUserId; // Protect the Cognito ID link
      delete userData.createdAt;

      // Update the timestamp
      userData.updatedAt = new Date();

      const result = await collection.findOneAndUpdate(
        { _id: userId }, // No need for ObjectId since we're using cognitoUserId as _id
        { $set: userData },
        { returnDocument: "after" }
      );

      return result;
    } catch (error) {
      throw new Error(`Error updating user: ${error.message}`);
    }
  }

  /**
   * Update user information by Cognito user ID
   * @param {string} cognitoUserId - Cognito user ID
   * @param {Object} userData - User data to update
   * @returns {Promise<Object|null>} Updated user object or null
   */
  async updateUserByCognitoId(cognitoUserId, userData) {
    try {
      const db = await connectToDatabase();
      const collection = db.collection(COLLECTION_NAME);

      // Don't allow updating these fields directly
      delete userData._id; // _id can't be updated
      delete userData.createdAt; // don't allow changing creation date

      // Update the timestamp
      userData.updatedAt = new Date();

      // Update or insert user document with cognitoUserId as _id
      const result = await collection.findOneAndUpdate(
        { _id: cognitoUserId },
        { $set: userData },
        { returnDocument: "after", upsert: true }
      );

      return result;
    } catch (error) {
      throw new Error(`Error updating user: ${error.message}`);
    }
  }
}

module.exports = new UserService();
